xbuild/src/command/doctor.rs

```diff
@@ -131,22 +154,27 @@ impl VersionCheck {
 }
 
 impl Check {
-    fn name(self) -> &'static str {
+    fn name(&self) -> &'static str {
         self.name
     }
 
-    fn path(self) -> Result<PathBuf> {
-        Ok(which::which(self.name)?)
+    fn path(&self) -> Result<PathBuf> {
+        Ok(match &self.location {
+            Some(Ok(path)) => path.clone(),
+            // Cannot clone the error:
+            Some(Err(e)) => bail!("{:?}", e),
+            None => which::which(self.name)?,
+        })
     }
 
-    fn version(self) -> Result<Option<String>> {
+    fn version(&self) -> Result<Option<String>> {
         if let Some(version) = self.version {
-            let output = Command::new(self.name)
+            let output = Command::new(self.path()?)
                 .args(version.arg.split(' '))
                 .output()?;
             anyhow::ensure!(output.status.success(), "failed to run {}", self.name);
             let output = std::str::from_utf8(&output.stdout)?;
-            if let Some(line) = output.split('\n').nth(version.row as _) {
+            if let Some(line) = output.lines().nth(version.row as _) {
```

Comment:
Drive-by fix (that I can PR separately): on Windows `adb --version` returns `\r\n`, so the `\r` remains in the `version()` string and clears the line (`adb     1.0.41\r`) that was printed before it, leaving only `path()` 🤭